specs/agents-versioning-publishing.md

```diff
@@ -0,0 +1,62 @@
+# Hypertrace Agents versioning and publishing
+
+This document describes versioning and publishing requirements for Hypertrace agents.
+All Hypertrace agents follow these requirements:
+
+* Semantic versioning for all released artifacts.
+* Release every merge commit into the main or maintenance branch as `"snapshot"`.
+* Maintenance branches for older versions should be named as `release-<major>.<minor>` e.g. `release-1.0`.
+
+#### Semantic versioning
+
+Each released version strictly follows [semantic versioning](https://semver.org/).
+The released version must be tagged in the source code: e.g. `1.0.0` or `v1.0.0` for languages
+with this convention (e.g. Golang).
+
+#### Snapshot release for every merge commit
+
+Each commit merged into the `main` or maintenance branch (e.g. `release-1.1` for `1.1.x`) should be
+published. The version of this release must include commit `SHA` and should include the last released
+semantic version and `snapshot` string to denote that the release should not be used in production environment.
+The last released semantic version helps to understand which version a given artifact is based on.
+The snapshot releases are not tagged in the source control and have no maintenance period.
+
+An example of the version:
+* `1.0.1-2a0d8cc26018b09763f5b1dcdf925b6f46ea1882`.
+* `1.0.1-SNAPSHOT-2a0d8cc26018b09763f5b1dcdf925b6f46ea1882`.
+
+##### Java
+
+The snapshot releases should not be published to the main
```

Comment:
It is strange to say something obvious like don't push snapshots to a release repository, especially not mentioning what is correct.

bintray's snapshot analog is jfrog, though I would suggest not advertising bintray at all officially.

Unlike sonatype releases repository or maven central, in bintray, your org can be locked based on consumption. For example, if someone downloads too much, you will not be able to do releases anymore until you adjust that or pay for something premium. Even if we already pay a premium, there are likely limits to consider.

Reply:
> bintray's snapshot analog is jfrog, though I would suggest not advertising bintray at all officially.

+1, on the other hand documenting this make it easier for consumers to find the artifacts.

Reply:
> It is strange to say something obvious like don't push snapshots to a release repository, especially not mentioning what is correct.

note that our terminology "snapshot" does not mean classic java `-SNAPSHOT`, hence I would like to make it obvious that these releases should be mixed with the normal release stream.